cfe_internal/update/update_policy.cf

```diff
@@ -80,6 +80,47 @@ bundle agent cfe_internal_update_policy
 
 }
 
+bundle agent cfe_internal_remove_broken_symlink(symlink_path)
+# @brief Checks if the symlink_path symlink points to an existing target
+# @warning Only works for symlinks using absolute paths!
+{
+  vars:
+      "link_target" string => filestat("$(symlink_path)", "linktarget");
+
+  files:
+      "$(symlink_path)"
+        delete => u_tidy,
+        if => not(fileexists("$(link_target)"));
+}
+
+bundle agent cfe_internal_setup_python_symlink(symlink_path)
+# @brief Create the /var/cfengine/bin/python symlink pointing to some installed python (if any)
+{
+  vars:
+      "path" string => getenv("PATH", 1024);
+      "path_folders" slist => splitstring("$(path)", ":", 128);
+
+      "exact_version_globs" slist => maplist("$(this)/python[23]", @(path_folders));
+      "generic_python_globs" slist => maplist("$(this)/python", @(path_folders));
+
+      "python_exact[$(exact_version_globs)]" slist => findfiles("$(exact_version_globs)");
+      "python_generic[$(generic_python_globs)]" slist => findfiles("$(generic_python_globs)");
+
+      "python_exact_sorted" slist => reverse(sort(getvalues(@(python_exact)), "lex")),
+        comment => "Prefer higher major versions of Python";
+
+      "pythons" slist => getvalues(mergedata(@(python_exact_sorted), getvalues(@(python_generic)))),
+        comment => "Prefer exact versions over unspecified";
+
+      "python" string => nth(@(pythons), 0),
+        if => isgreaterthan(length(@(pythons)), 0);
+
+  files:
+      "$(symlink_path)"
+        link_from => u_ln_s("$(python)"),
+        if => isgreaterthan(length(@(pythons)), 0);
```

Comment:
Instead of having a seperate bundle to remvoe the busted symlink, I suggest removing the symlink in the event there are no pythons.

Something like:

```
  files:
      # If we find no usable python versions, then let's remove the symlink instead of pointing to something that doesn't work.
      "$(symlink_path)"
        delete => u_tidy,
        if => not( isvariable( python ) );

      # If we find a usable python version, lets be sure the symlink points to the right place
      "$(symlink_path)"
        link_from => u_ln_s("$(python)"),
        move_obstructions => "true",
        if => isgreaterthan(length(@(pythons)), 0);
```

Reply:
How is this gonna help me if I uninstall Python 2 and install Python 3 instead? (between two agent runs).

Reply:
Well, the next agent run, the python variable would presumably point to pyhon3 instead of python2, then the promise that the symlink_path points to $(python) would be updated since it was previousy pointing to python2. Have I overlooked some detail where that would not be the case?

Reply:
Well, `python` will be a variable so the symlink won't get removed. And as I said, I don't exactly know what `move_obstructions` is doing -- I guess it would only change the symlink if the target differs?

Reply:
Here is a small example for you:

    bundle agent example_move_obstructions
    {
      files:
        "/tmp/symlink"
          link_from => ln_s( "/tmp/target" ),
          move_obstructions => "true";
    }
    bundle agent __main__
    {
      methods: "example_move_obstructions";
    }

Output:

    info: Linked files '/tmp/symlink' -> '/tmp/target'

Check the symlink:

    ls -al /tmp/symlink  

    lrwxrwxrwx 1 nickanderson nickanderson 11 Jun  6 11:27 /tmp/symlink -> /tmp/target

Subsequent execution produces no output

    bundle agent example_move_obstructions
    {
      files:
        "/tmp/symlink"
          link_from => ln_s( "/tmp/target" ),
          move_obstructions => "true";
    }
    bundle agent __main__
    {
      methods: "example_move_obstructions";
    }

Lets re-point the symlink.

    ln -sf /tmp/target2 /tmp/symlink  
    ls -al /tmp/symlink  

Output:

    lrwxrwxrwx 1 nickanderson nickanderson 12 Jun  6 11:29 /tmp/symlink -> /tmp/target2

And run the policy again:

    bundle agent example_move_obstructions
    {
      files:
        "/tmp/symlink"
          link_from => ln_s( "/tmp/target" ),
          move_obstructions => "true";
    }
    bundle agent __main__
    {
      methods: "example_move_obstructions";
    }

Output: 

    info: Overriding incorrect link '/tmp/symlink'
    info: Linked files '/tmp/symlink' -> '/tmp/target'

And replace the symlink with a directory

    rm -f /tmp/symlink  
    mkdir /tmp/symlink
    ls -al /tmp/symlink  

Output:

    total 4564
    drwxrwxr-x   2 nickanderson nickanderson    4096 Jun  6 11:32 .
    drwxrwxrwt+ 39 root         root         4665344 Jun  6 11:33 ..

And run the policy again:

    bundle agent example_move_obstructions
    {
      files:
        "/tmp/symlink"
          link_from => ln_s( "/tmp/target" ),
          move_obstructions => "true";
    }
    bundle agent __main__
    {
      methods: "example_move_obstructions";
    }

Output:

    info: Linked files '/tmp/symlink' -> '/tmp/target'

Again, with move\_obstructions, when it's in the correct state, the symlink is not re-written (no output).

    bundle agent example_move_obstructions
    {
      files:
        "/tmp/symlink"
          link_from => ln_s( "/tmp/target" ),
          move_obstructions => "true";
    }
    bundle agent __main__
    {
      methods: "example_move_obstructions";
    }

Reply:
The delete => tidy promise is not for the case of switching python versions, it's for the case where no python is found.

Reply:
OK, looks good.